### ISSUE TITLE
OSX large message corruption

### ISSUE THREAD

**Issue by HenrikPoulsen** [Azure/DotNetty#323]
Triggered on: macOS Sierra 10.12.6 (16G1114)
Dotnetty version: 0.4.7
Google.Protobuf: 3.5.0 

When using DotNetty for a project at work I started noticing when I was transferring large Protobuf messages between devices it was failing whenever it was sent from an OSX machine.
From linux and Windows I can send messages the size of 32MB without issues (I don't need larger than that so I haven't tested, but as for OSX I can only consistently send a couple of KB, larger than that and the messages end up garbled and cutoff on the receiving end.

I created a project consisting of some tests that show this:
https://github.com/HenrikPoulsen/DotNettyBug

<img width="784" alt="screen shot 2017-12-09 at 16 09 28" src="https://user-images.githubusercontent.com/472627/33796763-649c7202-dcfb-11e7-9f3d-a1d3a18e302f.png">

The tests transmit a simple protobuf message consisting of a random string of the sizes 32, 64, 128, 256, 512, 1024KB. As you can see from the screenshot when using EmbeddedChannel it is successful in all cases, but when using a regular tcp connection it starts failing somewhere between 128KB and 256KB. I have tried narrowing it down to a consistent value, but the threshold seems to vary depending on the receiving system (a mac recipient seem to be able to handle larger messages than a windows one)

If I run the same tests from the repo posted above on a Windows 10 machine they are all successful.

For now my workaround is that all data I am sending for windows and linux have the regular 32MB cap, but if OSX is sending it I have set it down to a measly 1KB to make sure the issue isn't triggered.

If I remove the lower send limit of my OSX machine it gives the following exceptions on the receiving side:
``` C#
Exception: DotNetty.Codecs.DecoderException: Exception of type 'DotNetty.Codecs.DecoderException' was thrown. ---> DotNetty.Codecs.CodecException: Exception of type 'DotNetty.Codecs.CodecException' was thrown. ---> Google.Protobuf.InvalidProtocolBufferException: SkipLastField called on an end-group tag, indicating that the corresponding start-group was missing
   at Google.Protobuf.CodedInputStream.SkipLastField()
   at Unity.DistributedTestRunner.Core.GeneratedMessage.MessageWrapper.MergeFrom(CodedInputStream input) in /Users/henrikpoulsen/distributed-test-runner/Unity.DistributedTestRunner.Core/GeneratedMessage/MessageWrapper.cs:line 168
   at Google.Protobuf.MessageParser.ParseFrom(CodedInputStream input)
   at DotNetty.Codecs.Protobuf.ProtobufDecoder.Decode(IChannelHandlerContext context, IByteBuffer message, List`1 output)
   --- End of inner exception stack trace ---
   at DotNetty.Codecs.Protobuf.ProtobufDecoder.Decode(IChannelHandlerContext context, IByteBuffer message, List`1 output)
   at DotNetty.Codecs.MessageToMessageDecoder`1.ChannelRead(IChannelHandlerContext context, Object message)
   --- End of inner exception stack trace ---
   at DotNetty.Codecs.MessageToMessageDecoder`1.ChannelRead(IChannelHandlerContext context, Object message)
   at DotNetty.Transport.Channels.AbstractChannelHandlerContext.InvokeChannelRead(Object msg)
```
This is however not triggered until after my MessageHandler has received the message and the content has been seemingly corrupted (a large section of the message has been repeated and the end has been cut off)
Below you can see a screenshot of a kdiff3 view of the text file I used for this sample, on the left is the original and the right is after it has been received.
<img width="980" alt="screen shot 2017-12-09 at 16 23 57" src="https://user-images.githubusercontent.com/472627/33796869-70eacf98-dcfd-11e7-8788-4a29a61b1968.png">
So part ways into the message the beginning of the message is repeating itself, and at the bottom there is a whole section missing from the received file.

I hope this report is useful. You don't want to know how much time I have spent staring at this issue now :)

Edit:
Added some version information up top

**Comment by caozhiyuan** [comment]
look this https://github.com/Azure/DotNetty/pull/299, it also has bug in .netcore 2.0.3 . change code to  
  caozhiyuan:linuxsocketbugfix and then try

**Comment by HenrikPoulsen** [comment]
@caozhiyuan Your branch does in fact fix this issue for me on OSX. I modified my tests to send up to 32MB and as you can see in the image below it is working.
<img width="653" alt="screen shot 2017-12-10 at 13 11 38" src="https://user-images.githubusercontent.com/472627/33804795-d97301ca-ddac-11e7-908e-dd015ca7f71a.png">
Fingers crossed your changes land soon :)

**Comment by nayato** [comment]
should be fixed with #336
